// GenericAccordion/GenericAccordion.js
import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { styled } from '@mui/material/styles/index.js';

import BaseAccordion from '@mui/material/Accordion/index.js';
import AccordionSummary from '@mui/material/AccordionSummary/index.js';
import AccordionDetails from '@mui/material/AccordionDetails/index.js';

import ExpandMoreIcon from '@mui/icons-material/ExpandMore.js';

import { useLocalStorage } from 'react-use';

const Accordion = styled((props) => (
    <BaseAccordion disableGutters elevation={0} square {...props} />
))(() => ({
    border: `1px solid #b7b7b7`,
    backgroundColor: '#e7e7e7'
}));

const GenericAccordion = function ({
    children,
    localStorageIdForExpanded,
    localStorageDefaultValueForExpanded,
    title,
    icon
}) {
    let expanded;
    let setExpanded;
    if (localStorageIdForExpanded) {
        [expanded, setExpanded] = useLocalStorage(localStorageIdForExpanded, localStorageDefaultValueForExpanded || 'no');
    } else {
        [expanded, setExpanded] = useState('no');
    }
    const [initialized, setInitialized] = useState(expanded === 'yes');

    return (
        <Accordion
            TransitionProps={{
                unmountOnExit: initialized ? false : true
            }}
            expanded={expanded === 'yes'}
            onChange={(evt, expanded) => {
                if (expanded) {
                    setInitialized(true);
                }

                setExpanded(expanded ? 'yes' : 'no');
            }}
        >
            <AccordionSummary expandIcon={<ExpandMoreIcon />}>
                {
                    icon &&
                    <div>
                        {icon}
                    </div>
                }
                <div
                    style={{
                        marginLeft: (icon ? 10 : 0)
                    }}
                >
                    {title}
                </div>
            </AccordionSummary>
            <AccordionDetails
                style={{
                    backgroundColor: '#fff',
                    borderTop: '1px solid #bbb'
                }}
            >
                <div
                    style={{
                        paddingTop: 8,
                        width: '100%'
                    }}
                >
                    {children}
                </div>
            </AccordionDetails>
        </Accordion>
    );
};
GenericAccordion.propTypes = {
    children: PropTypes.node.isRequired,
    localStorageIdForExpanded: PropTypes.string,
    localStorageDefaultValueForExpanded: PropTypes.string,
    title: PropTypes.node.isRequired,
    icon: PropTypes.node
};

export { GenericAccordion };
